sklearn/pipeline.py

```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
         for t in transformers:
             if t is None or t == 'passthrough':
                 continue
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
-                    hasattr(t, "transform")):
+            if (not hasattr(t, "fit_transform") and
+                    (not (hasattr(t, "fit") and hasattr(t, "transform")))):
                 raise TypeError("All intermediate steps should be "
-                                "transformers and implement fit and transform "
-                                "or be the string 'passthrough' "
-                                "'%s' (type %s) doesn't" % (t, type(t)))
+                                "transformers and implement fit and "
+                                "transform, fit_transform or be the string "
+                                "'passthrough'. '%s' (type %s) doesn't"
+                                % (t, type(t)))
+            elif not hasattr(t, "transform"):
+                warnings.warn("Intermediate step '%s' (type %s) does not have "
+                              "transform, pipeline is not reusable."
+                              % (t, type(t)))
 
         # We allow last estimator to be None as an identity transformation
         if (estimator is not None and estimator != 'passthrough'
-                and not hasattr(estimator, "fit")):
+                and not (hasattr(estimator, "fit") or
+                         hasattr(estimator, "fit_transform"))):
             raise TypeError(
-                "Last step of Pipeline should implement fit "
-                "or be the string 'passthrough'. "
+                "Last step of Pipeline should implement fit, "
+                "fit_transform or be the string 'passthrough'. "
                 "'%s' (type %s) doesn't" % (estimator, type(estimator)))
+        elif not (hasattr(estimator, "transform") or
```

Comment:
This is not an intermediate step. We do not need to handle the case you've handled here.

Reply:
Don't you think it is worth raising a warning in that case? Or the user should expect this since it's the final step of the pipeline, and the existing code (that checks which methods are available for the last step) is enough?

Reply:
Personally, I would find it more elegant to remove that code and leave it up to the user.  If somebody throws something like `TSNE` in there, I would expect that person to know what to do.